bin/fasta-identity-table.py

```diff
@@ -0,0 +1,365 @@
+#!/usr/bin/env python
+
+from __future__ import print_function, division
+
+import argparse
+from collections import OrderedDict, defaultdict
+
+from dark.reads import addFASTACommandLineOptions, parseFASTACommandLineOptions
+from dark.dna import compareDNAReads
+
+
+def explanation(matchAmbiguous, concise, showLengths):
+    """
+    Make an explanation of the output HTML table.
+
+    @param matchAmbiguous: If C{True}, count ambiguous nucleotides that are
+        possibly correct as actually being correct. Otherwise, we are strict
+        and insist that only non-ambiguous nucleotides can contribute to the
+        matching nucleotide count.
+    @param concise: If C{True}, do not show match detail abbreviations.
+    @param showLengths: If C{True}, include the lengths of sequences.
+    @return: A C{str} of HTML.
+    """
+    result = ["""
+<h1>Sequence versus sequence identity table</h1>
+
+<p>
+
+The table cells below show the nucleotide identity fraction for the sequences
+(<span class="best">like this</span> for the best value in each row). The
+identity fraction numerator is the sum of the number of identical
+    """]
+
+    if matchAmbiguous:
+        result.append('nucleotides, plus the number of ambiguously matching '
+                      'nucleotides, ')
+    else:
+        result.append('nucleotides')
+
+    result.append("""
+plus the number of sites where both sequences have gaps. The denominator
+is the length of the sequence <em>for the row</em>.
+
+</p>
+    """)
+
+    if showLengths or matchAmbiguous or not concise:
+        result.append("""
+<p>
+
+Key to abbreviations:
+  <ul>
+    """)
+
+        if showLengths:
+            result.append('<li>L: sequence Length.</li>')
+
+        if not concise:
+            result.append('<li>IM: Identical nucleotide Matches.</li>')
+
+        if matchAmbiguous:
+            result.append('<li>AM: Ambiguous nucleotide Matches.</li>')
+
+        result.append("""
+    <li>GG: Gap/Gap matches (both sequences have gaps).</li>
+    <li>G?: Gap/Non-gap mismatches (one sequence has a gap).</li>
+    <li>NE: Non-equal nucleotide mismatches.</li>
+  </ul>
+</p>
+""")
+
+    return '\n'.join(result)
+
+
+def collectData(reads1, reads2, square, matchAmbiguous):
+    """
+    Get pairwise matching statistics for two sets of reads.
+
+    @param reads1: An C{OrderedDict} of C{str} read ids whose values are
+        C{Read} instances. These will be the rows of the table.
```

Comment:
Do you need to make an OrderedDict? Can you not just pass the Reads instance as is? Or is that not ordered?

Reply:
I could have just used the `Reads` but it's a bit awkward to look things up in because it's neither a `dict` nor a `list`. The other option would be to use `ReadsInRAM` which is a bit like a list. Anyway, I went for an OrderedDict so I can deal with sequence ids and keep things in the order they appear in the FASTA file(s).